### ISSUE TITLE
ArgoCD hook resources are randomly deleted after the `PreSync` hook finishes

### ISSUE THREAD

**Issue by rodrigoechaide** [argoproj/argo-cd#13429]
Checklist:

* [X] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [X] I've included steps to reproduce the bug.
* [X] I've pasted the output of the `argocd version`.

**Describe the bug**

We are experiencing a weird issue using ArgoCD. We have migrated an application we installed into the k8s cluster using Helm to ArgoCD. In this app, we use argocd hooks in some `configMaps`, `secretProviderClass`, and `job` objects to perform database migrations. After the migration to argo we are having the issue that the `configMaps` and the `secretProviderClass` objects are randomly deleted after the syncs but not on all the syncs (we are having several syncs a day), only in some of them. To give you more details these objects with hook configurations have the hook policy `BeforeHookCreation` so before each sync they are recreated and after the PreSync hook has finished (more specifically the job that runs the database migrations has finished) the configMaps and secretProviderClass objects are randomly deleted. After this, we have to run a manual sync (pressing the sync button in argocd UI) so the `PreSync` hook is invoked again and all the resources in the hook are created again and hoping that they are not deleted after the `PreSync` hook finishes.

**Note**: Right after the migration from helm to argo we were still using helm annotations for the hooks (it is supposed that ArgoCD translates the annotation into ArgoCD hooks annotations) and we experienced the same issue. Yesterday we changed the annotations to the [argocd ones](https://argo-cd.readthedocs.io/en/stable/user-guide/resource_hooks/) and we are experiencing the same issue.

**To Reproduce**

* Create a release with argo using helm as the template engine and in some of the objects (`Job` and `configMaps`) put the following annotations:
  * For the Job:
    * argocd.argoproj.io/hook: PreSync
    * argocd.argoproj.io/hook-delete-policy: BeforeHookCreation
    * argocd.argoproj.io/sync-wave: 0
  * For the configMap:
    * argocd.argoproj.io/hook: PreSync
    * argocd.argoproj.io/hook-delete-policy: BeforeHookCreation
    * argocd.argoproj.io/sync-wave: -5
* After several argoCD syncs the configMaps and job objects should be randomly deleted after the `PreSync` hook finishes.

**Expected behavior**

After the PreSync hook finishes all the resources created by the hook are not deleted and are still there until the next sync when the `PreSync` hook is invoked again and all the resources with the hook are recreated.

**Screenshots**

No Screenshots

**Version**

```shell
argocd-server: v2.5.0+b895da4
```

**Logs**

```
No relevant logs.
```

**Comment by rodrigoechaide** [comment]
@bcbrockway In our case, we had to duplicate `configMaps` and `secret` objects for the pre-sync phase and for the main sync phase so the ones in the main sync phase are always there and not deleted.

**Comment by davhdavh** [comment]
This killed one of our nodes yesterday when it got stuck in a fail/job retry/delete/recreate loop, that ended up with etcd getting OOM killed.

**Comment by bcbrockway** [comment]
> @bcbrockway In our case, we had to duplicate the `configMaps` and `secretProviderClass ` objects for the pre-sync phase and for the main sync phase so the ones in the main sync phase are always there and not deleted.

Yeah that's what we're going to end up doing I think.

**Comment by PabloB94** [comment]
This is affecting us as well, any idea when this will be worked on?

**Comment by alexvaigm** [comment]
Hi, any updates here ? We facing the same in version v2.11.2

**Comment by andrii-korotkov-verkada** [comment]
Out of curiosity, what's the reason to make some config maps to be hooks? Hook is more intuitively associated with some kind of a job to me, while for persistent objects it seems better to use sync wave -1 or so.

**Comment by davhdavh** [comment]
> Out of curiosity, what's the reason to make some config maps to be hooks? Hook is more intuitively associated with some kind of a job to me, while for persistent objects it seems better to use sync wave -1 or so.

That you can have some config specifically for the hook that doesn't need to stay around.

**Comment by rodrigoechaide** [comment]
This issue is still happening for us even with argo `v2.11.0+d3f33c0`

**Comment by andrii-korotkov-verkada** [comment]
Why do you need a config map as a hook? These should normally be managed using sync waves.

**Comment by bcbrockway** [comment]
Because sometimes you need a ConfigMap to configure a workload that runs as a hook.

**Comment by andrii-korotkov-verkada** [comment]
If the hook resource is needed for non-hooks as well, I suggest using sync waves. E.g. create a CM on sync wave -1, run hook on resource with sync wave 0.

**Comment by yafeiaa** [comment]
Hi, is there any progress in this? We meet this problem at v2.9.0 too.

**Comment by Sturgelose** [comment]
Same issue here and solutions (and failures I've had):

1. Try @andrii-korotkov-verkada proposal of using syncwaves + hooks. It simply doesn't work. Each hook has its own waves. So the order is: presync -1, presync 0, presync 1, sync -1, sync 0, sync 1. If you want to have a configmap in a presync, you are forced to set those resources as presync hooks as well, and... they disappear randomly

2. Do not use presync hooks at all and just use waves. This is not sometimes available because we cannot override some helm charts. ~~But if you even try it, and you define a job in a web without a hook, the job gets created, you need to configure with replace/force to make sure you recreate it. And once the job expires, sync will kick it again and run it again. So it kinda becomes like an unwanted cronjob...~~ Update: figured it out the TTL was cleaning up the job, unsetting it seems to fix the issue!

3. Define the configmap twice. Seems it is the only working solution. The bad thing is that this means duplicating configuration, resources and extra complexity in kustomizations as a workaround.

And not to say, this problem gets extended when you try to use sealedSecrets or ExternalSecrets. You want the secrets to be created before the pods consume them!

The real solution would be either:
1.  Make sure resources do not get deleted/recreated in preSync. Including non primitive resources such as ExternalSecrets. Maybe a flag to tell so would be useful!
2. Have a specific "hook", before everything else that inits secrets and configmaps and anything else you can annotate/indicate, and make sure you only clean this up at the end of a successful sync. Otherwise you can end with pods having their current configmap deleted.

**Comment by morganbasset** [comment]
Why the resources in pre-sync are not kept if "hook-delete-policy" is not defined ? 
As said before, something your migration database job need external secret like "vault" mounted with kubernetes CRD, you also need RBAC that are the same used by your deployment in sync phase.
Please bring the possibility to keep resource define in pre-sync.

**Comment by hanikesn** [comment]
> Please bring the possibility to keep resource define in pre-sync.

Yes, just hit the issue today again, service accounts, configmaps, secret stores and e.g. client certs all need to be defined twice when using pre-sync hooks.

**Comment by xtimon** [comment]
Is there any chance to resolve this issue?
